### PR TITLE
Fix race condition in servodriver

### LINKED ISSUE #16754: Possible race condition in get_free_port
I am experimenting with WebDriver in Servo and came across with what might be a race condition in the [`get_free_port`](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/browsers/base.py#L39) function. That is, sometimes when `wptrunner` runs the tests in parallel, two webdriver instances get assigned the same port.
You can find all the details here: https://github.com/servo/servo/issues/22619#issuecomment-490587018.

### PR DESCRIPTION
**Pull request by georgeroman**
Fixes #16754

### CODE REVIEW

File: tools/wptrunner/wptrunner/browsers/servodriver.py [hunk 1]
```diff
@@ -94,8 +95,12 @@ def __init__(self, logger, binary, debug_info=None, webdriver_host="127.0.0.1",
...
 
     def start(self, **kwargs):
-        self.webdriver_port = get_free_port(4444, exclude=self.used_ports)
-        self.used_ports.add(self.webdriver_port)
+        ServoWebDriverBrowser.used_ports_lock.acquire()
+        try:
+            self.webdriver_port = get_free_port(4444, exclude=self.used_ports)
+            self.used_ports.add(self.webdriver_port)
+        finally:
+            ServoWebDriverBrowser.used_ports_lock.release()
```
Comment (jdm) [inline tools/wptrunner/wptrunner/browsers/servodriver.py]: Let's do this instead:
```python
with ServoWebDriverBrowser.used_ports_lock:
    self.webdriver_port = get_free_port(4444, exclude=self.used_ports)
    self.used_ports.add(self.webdriver_port)
```

**Review by jdm — CHANGES_REQUESTED** [review]

File: tools/wptrunner/wptrunner/browsers/servodriver.py (hunk 1 above)
Comment (jugglinmike) [reply tools/wptrunner/wptrunner/browsers/servodriver.py]: You beat me to it!

**Review by jdm — APPROVED** [review]

**Comment by gsnedders** [comment]
I wonder if every usage of `get_free_port` is racy in the face of parallel test execution?

**Comment by georgeroman** [comment]
I think it is since every usage of the function follows the same pattern of getting the first available port followed by saving the chosen port in `excluded_ports`.

**Comment by SimonSapin** [comment]
Binding to port 0 lets the OS pick an available port at semi-random, without the race. `getsockname` can then be used to find out which port was used.

**Comment by jgraham** [comment]
I seem to recall that there's a problem doing that, at least with the way that marionette works, because we need to pass the port number in to the command line argument rather than bind it ourselves. So we *could* get a socket with a random/unused port then read it's port number and then pass that down after closing it but we can't get a socket on a random port and then use that socket directly.

**Pull request merged**